elixir/lib/contracts.ex

```diff
@@ -25,36 +81,68 @@ defmodule Trento.Contracts do
         DateTime.utc_now()
       )
 
+    expiration = DateTime.add(time, validity_in_seconds)
+
     time_attr = %Google.Protobuf.Timestamp{seconds: time |> DateTime.to_unix()}
+    expire_at_attr = %Google.Protobuf.Timestamp{seconds: expiration |> DateTime.to_unix()}
+
     data = Protobuf.Encoder.encode(struct)
 
-    cloud_event = %CloudEvents.CloudEvent{
+    %CloudEvents.CloudEvent{
       data: {:proto_data, %Google.Protobuf.Any{value: data, type_url: get_type(mod)}},
       spec_version: "1.0",
       type: get_type(mod),
       id: id,
       attributes: %{
-        "time" => %CloudEvents.CloudEventAttributeValue{attr: {:ce_timestamp, time_attr}}
+        "time" => %CloudEvents.CloudEventAttributeValue{attr: {:ce_timestamp, time_attr}},
+        "expire_at" => %CloudEvents.CloudEventAttributeValue{
+          attr: {:ce_timestamp, expire_at_attr}
+        }
       },
       source: source
     }
+  end
+
+  defp add_signature(
+         %CloudEvents.CloudEvent{
+           data: {:proto_data, %Google.Protobuf.Any{value: data}},
+           attributes:
+             %{
+               "time" => %CloudEvents.CloudEventAttributeValue{
+                 attr: {:ce_timestamp, %{seconds: time}}
+               },
+               "expire_at" => %CloudEvents.CloudEventAttributeValue{
+                 attr: {:ce_timestamp, %{seconds: expire_at}}
+               }
+             } = current_attrs
+         } = event,
+         private_key_content
+       ) do
+    signing_key =
+      private_key_content
+      |> :public_key.pem_decode()
+      |> Enum.at(0)
+      |> :public_key.pem_entry_decode()
 
-    CloudEvents.CloudEvent.encode(cloud_event)
+    time_str = Integer.to_string(time)
+    signature = :public_key.sign("#{data}#{time_str}#{expire_at}", :sha256, signing_key)
```

Comment:
[I don't think protobuf serde is canonical](https://protobuf.dev/programming-guides/serialization-not-canonical/). See for instance this [RFC as an example of canonicalization over JSON](https://www.rfc-editor.org/rfc/rfc8785). Signing+verification (and many cryptographic operations) over non-canonical structures are not reliably repeatable, if I am understanding this correctly. WDYT?